File: COMPILADOR/kokoslan_/src/src/java/ast/KoKoLet.java
/**
 * @author loriacarlos@gmail.com
 *
 */
package src.java.ast;

import java.io.*;
import src.java.eval.KoKoContext;
import src.java.eval.KoKoValue;

public class KoKoLet implements KoKoAst {

    private KoKoAst id;

    public KoKoAst getId() {
        return this.id;
    }

    private KoKoAst expr;

    public KoKoAst getExpr() {
        return this.expr;
    }

    public KoKoLet(KoKoAst id, KoKoAst expr) {
        this.id = id;
        this.expr = expr;
    }

    public void genCode(PrintStream out) {
        out.print("let ");
        this.id.genCode(out);
        out.print(" = ");
        this.expr.genCode(out);
        out.println();
    }

    public KoKoValue eval(KoKoContext ctx) {
        KoKoValue val = expr.eval(ctx);
        ctx.assoc((KoKoId) id, (KoKoValue) val);
        return val;
    }
}

File: COMPILADOR/kokoslan/src/KoKoKotlin/ast/KoKoLet.kt
/**
 * @author loriacarlos@gmail.com
 */
package KoKoKotlin.ast

import KoKoKotlin.eval.KoKoContext
import KoKoKotlin.eval.KoKoValue
import java.io.*

import src.java.ast.KoKoAst


class KoKoLet(val id: KoKoAst, val expr: KoKoAst) : KoKoAst {

    override fun genCode(out: PrintStream) {
        out.print("let ")
        this.id.genCode(out)
        out.print(" = ")
        this.expr.genCode(out)
        out.println()
    }

    override fun eval(ctx: KoKoContext): KoKoValue {
        val `val` = expr.eval(ctx)
        ctx.assoc(id as KoKoId, `val` as KoKoValue)
        return `val`
    }
}
